### PR TITLE
Update map_epiflows

### PR DESCRIPTION
**Pull request by zkamvar**
I've updated `map_epiflows()` so that it can take an epiflows object with coordinates and map it using leaflet. I changed the colors, but they can be easily changed back. It still needs a bit of polishing, but as it is now, it works.

Other things I did to get things working:

 - added `get_coords()` to return a table of coordinates with IDs OR a vector for a single location.
 - modified `add_coordinates()` for the new data format
 - added a validator for flows that checks to make sure there are no duplicates
 - make it so the `get_vars()` function returns the `id` column with its result

Some things I still need to do:

 - add a `merge()` call into `add_coordinates()` to allow for a data frame with different row orders.
 - tests tests tests
 - document a bug I found with leaflet where named lists internally cause the lines not to appear

The example should produce the following:

``` r
library("epiflows")
library("tibble")
location_df <- tibble::tribble(
  ~id,        ~lon,       ~lat,
  "ARG",  -63.616672, -38.416097,
  "BEL",    4.469936,  50.503887,
  "BHS",   -77.39628,   25.03428,
  "BLZ",   -88.49765,  17.189877,
  "BOL",  -63.588653, -16.290154,
  "BRA",   -51.92528, -14.235004,
  "BRB",  -59.543198,  13.193887,
  "CAN", -106.346771,  56.130366,
  "CHE",    8.227512,  46.818188,
  "CHL",  -71.542969, -35.675147,
  "CHN",  104.195397,   35.86166,
  "COL",  -74.297333,   4.570868,
  "CRI",  -83.753428,   9.748917,
  "CUB",  -77.781167,  21.521757,
  "DEU",   10.451526,  51.165691,
  "DOM",  -70.162651,  18.735693,
  "ECU",  -78.183406,  -1.831239,
  "ESP",    -3.74922,  40.463667,
  "FIN",  25.7481511,   61.92411,
  "FRA",    2.213749,  46.227638,
  "GBR",   -3.435973,  55.378051,
  "GTM",  -90.230759,  15.783471,
  "HKG",  114.109497,  22.396428,
  "HND",  -86.241905,  15.199999,
  "ITA",    12.56738,   41.87194,
  "JAM",  -77.297508,  18.109581,
  "JPN",  138.252924,  36.204824,
  "LUX",    6.129583,  49.815273,
  "MEX", -102.552784,  23.634501,
  "MMR",   95.955974,  21.916221,
  "NIC",  -85.207229,  12.865416,
  "NLD",    5.291266,  52.132633,
  "PAN",  -80.782127,   8.537981,
  "PER",  -75.015152,  -9.189967,
  "PRT",   -8.224454,  39.399872,
  "PRY",  -58.443832, -23.442503,
  "SLV",   -88.89653,  13.794185,
  "SWE",   18.643501,  60.128161,
  "TTO",  -61.222503,  10.691803,
  "USA",  -95.712891,   37.09024,
  "VEN",   -66.58973,    6.42375
)
data("Mex_travel_2009")
flows   <- Mex_travel_2009[[1]]
outflow <- setNames(flows[["MEX"]], rownames(flows))
inflow  <- unlist(flows["MEX", , drop = TRUE])
ef      <- epiflows(inflow, outflow, focus = "MEX", locations = Mex_travel_2009[[2]])
#> Warning: Pruning duplicate entries in the flows.
ef      <- add_coordinates(ef, location_df[-1])
map_epiflows(ef)
```

image

Created on 2018-07-09 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

### CODE REVIEW

**Comment by zkamvar** [comment]
Update: 

I've also added an `as.SpatialLinesDataFrame()` method that is publicly-facing

File: DESCRIPTION [hunk 1]
```diff
@@ -29,5 +28,5 @@ Suggests:
     outbreaks
 RoxygenNote: 6.0.1.9000
 URL: https://github.com/reconhub/epiflows
-BugReports: https://github.com/reconhub/epimaps/issues
+BugReports: https://github.com/reconhub/epiflows/issues
```
Comment (thibautjombart) [inline DESCRIPTION:33]: Once we have a first stable version, we'll need to add a website (`pkgdown::build_site()`)

File: NAMESPACE [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
 S3method(get_locations,epiflows)
 S3method(get_vars,epiflows)
 S3method(plot,epiflows)
 S3method(print,epiflows)
 export(add_coordinates)
+export(as.SpatialLinesDataFrame)
 export(epiflows)
 export(estimate_risk_spread)
 export(get_codes)
+export(get_coords)
```
Comment (thibautjombart) [inline NAMESPACE]: change to `get_coordinates()` for name consistency with `add_coordinates()`

File: R/add_coordinates.R [hunk 3]
```diff
@@ -1,63 +1,88 @@
...
+    }
+    if (overwrite && !is.null(ef$vars$coords)) {
+      names(coords)               <- ef$vars$coords
+      ef$linelist[ef$vars$coords] <- coords
+    } else {
+      ef$linelist    <- cbind(get_locations(ef), coords) 
+      ef$vars$coords <- setNames(names(coords), c("lon", "lat"))
+    }
+  } else {
+    if (!is.character(coords) || length(coords) != 2) {
```
Comment (thibautjombart) [inline R/add_coordinates.R]: I increasingly assume I don't know how operator priority works, as I once had a nightmare juggling between R and C++; so I'd write   
```
    if (!is.character(coords) || (length(coords) != 2L)) {
```
but your call

File: R/get_vars.R [hunk 4]
```diff
@@ -3,8 +3,8 @@
 #' This accessor extracts variables from the `locations` data frame in an
 #' `epiflow` object. `get_vars` is a generic with a method defined for
 #' `epiflows` objects.
-#' 
-#' The 
+#'
+#' The function `get_coords()` is equivalent to `get_vars(x, "coords", id = TRUE)`
```
Comment (thibautjombart) [inline R/get_vars.R]: Maybe use `coordinates` throughout?

File: R/map_epiflows.R [hunk 5]
```diff
@@ -1,153 +1,125 @@
...
-  if (!all(lon_lat_columns %in% location_cols)) {
-    ## Tries to find the coordinates with add_coordinates(), should only do this if type == "map"
-    if(type=="map"){
-     x %<>%
-      add_coordinates(
-        loc_column = loc_column,
-        lon_lat_columns = lon_lat_columns
-      )
+  # Leaflet plot construction ---------------------------------------
+  pal    <- leaflet::colorQuantile(palette = "Greys",
```
Comment (thibautjombart) [inline R/map_epiflows.R]: I don't mind the grey, but people typically like pretty colors; maybe default to something more colorful, and pass it through as an argument to `map_epiflows(pal = "whatever")`; depending on how many palettes are available, maybe list them in the argument and use `match.arg`? Not super hard set on this, happy to chat

**Review by thibautjombart — CHANGES_REQUESTED** [review]

File: R/get_vars.R (hunk 4 above)
Comment (zkamvar) [reply R/get_vars.R]: Agreed. Done

File: R/add_coordinates.R (hunk 3 above)
Comment (zkamvar) [reply R/add_coordinates.R]: This scheme will work. I guarantee it. We can talk about your nightmares later.

File: NAMESPACE (hunk 2 above)
Comment (zkamvar) [reply NAMESPACE]: Done.

File: DESCRIPTION (hunk 1 above)
Comment (zkamvar) [reply DESCRIPTION:33]: Yes, and we will need to make an issue out of this :)

File: R/map_epiflows.R (hunk 5 above)
Comment (zkamvar) [reply R/map_epiflows.R]: I've changed the color palette to a yellow-orange-brown palette and also added line thickness which is `ln(n + 1)`

**Comment by zkamvar** [comment]
Here's what the map looks like with line weights (default):

image


Here's what the map looks like without line weights:

image

**Review by thibautjombart — APPROVED** [review]

**Comment by thibautjombart** [comment]
Great, thanks a lot!

**Pull request merged**